File: day1/day1.py
#!/bin/python
import re
from collections import Counter

def unzip(xs):
    return zip(*xs)

def part1(pairs):
    return sum(map(lambda x: abs(x[1] - x[0]), zip(*map(sorted, unzip(pairs)))))

def part2(pairs):
    xs, ys = unzip(pairs)
    counts = Counter(ys)
    return sum(x * (0 if x not in counts else counts.get(x)) for x in xs)

with open('input', 'r') as f:
    lines = f.read().splitlines()
    pairs = [tuple(map(int, re.findall(r'\d+', x))) for x in lines]
    part1 = part1(pairs)
    part2 = part2(pairs)
    print(part1, part2)

File: src/day1.ts
import { countBy, identity, sum, unzip, zipWith } from '@es-toolkit/es-toolkit'
import { findAll, print } from './utils.ts'

function part1(pairs: [number, number][]): number {
  const [xs, ys] = unzip(pairs)
    .map((x) => x.toSorted())
  return sum(zipWith(xs, ys, (a, b) => Math.abs(b - a)))
}

function part2(pairs: [number, number][]) {
  const [xs, ys] = unzip(pairs)
  const counts = countBy(ys, identity)
  return sum(xs.map((x) => x * (counts[x] ?? 0)))
}

export function runDay() {
  const parsed = Deno.readTextFileSync('./txt/day1')
    .split('\n')
    .filter((x) => x !== '')
    .map(findAll(/\d+/g))
    .map((x) => x.map(Number) as [number, number])
  const one = part1(parsed)
  const two = part2(parsed)
  print(one, two)
}
